storage/reads/merge.go

```diff
@@ -122,42 +122,81 @@ func (h *resultSetHeap) Pop() interface{} {
 	return item
 }
 
+// MergedStringIterator merges multiple storage.StringIterators into one.
+// It sorts and deduplicates adjacent values, so the output is sorted iff all inputs are sorted.
+// If all inputs are not sorted, then output order and deduplication are undefined and unpleasant.
 type MergedStringIterator struct {
-	iterators    []storage.StringIterator
-	uniqueValues map[string]struct{}
-	nextValue    string
+	heap      *stringIteratorHeap
```

Comment:
You can embed this to avoid an additional allocation

```suggestion
	heap      stringIteratorHeap
```